### PR TITLE
Refactor/merge

### PR DESCRIPTION
**Pull request by sixFingers**

### CODE REVIEW

File: src/inc/builder/sections/columns/definition.php [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+			get_template_directory() . '/inc/builder/',
+			$this->get_settings()
+		);
+
+		add_filter( 'make_section_defaults', array( $this, 'section_defaults' ) );
+		add_filter( 'make_get_section_json', array ( $this, 'get_section_json' ), 10, 2 );
+		add_filter( 'make_builder_js_dependencies', array( $this, 'add_js_dependencies' ) );
+	}
+
+	public function get_settings() {
```
Comment (tomasstark) [inline src/inc/builder/sections/columns/definition.php]: Consider doc block

File: src/inc/builder/sections/gallery/definition.php [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+				'type'    => 'color',
+				'label'   => __( 'Background color', 'make' ),
+				'name'    => 'background-color',
+				'class'   => 'ttfmake-gallery-background-color ttfmake-configuration-color-picker',
+				'default' => ttfmake_get_section_default( 'background-color', 'gallery' )
+			)
+		);
+	}
+
+	public function get_item_settings() {
```
Comment (tomasstark) [inline src/inc/builder/sections/gallery/definition.php]: Consider doc block

File: src/inc/builder/sections/gallery/definition.php [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+				}
+
+				array_push( $clean_data['gallery-items'], $clean_item_data );
+			}
+		}
+
+		return $clean_data;
+	}
+
+	public function add_js_dependencies( $deps ) {
```
Comment (tomasstark) [inline src/inc/builder/sections/gallery/definition.php]: Copy doc block from columns

File: src/inc/builder/sections/gallery/definition.php [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+			'captions' => 'reveal',
+			'caption-color' => 'light',
+			'background-image' => '',
+			'darken' => 0,
+			'background-style' => 'tile',
+			'background-color' => ''
+		);
+	}
+
+	public function get_item_defaults() {
```
Comment (tomasstark) [inline src/inc/builder/sections/gallery/definition.php]: Consider doc block

File: src/inc/builder/sections/gallery/definition.php [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+			),
+		) );
+
+		// Sort the config in case 3rd party code added another input
+		ksort( $inputs, SORT_NUMERIC );
+
+		return $inputs;
+	}
+
+	public function get_defaults() {
```
Comment (tomasstark) [inline src/inc/builder/sections/gallery/definition.php]: Copy doc block from columns

File: src/inc/builder/core/api.php [hunk 6]
```diff
@@ -75,25 +75,17 @@ public function get_sections() {
 	 * @param  string    $description         Section description.
 	 * @param  string    $icon                URL to the icon for the display.
 	 * @param  string    $save_callback       Function to save the content.
-	 * @param  string    $builder_template    Path to the template used in the builder.
+	 * @param  array     $builder_template    A path or array (section[, item]) of paths to the template(s) used in the builder.
```
Comment (sixFingers) [inline src/inc/builder/core/api.php]: Allow an hash of paths to be passed instead of a single path.

File: src/inc/builder/core/api.php [hunk 7]
```diff
@@ -75,25 +75,17 @@ public function get_sections() {
...
 	 * @param  string    $display_template    Path to the template used for the frontend.
 	 * @param  int       $order               The order in which to display the item.
 	 * @param  string    $path                The path to the template files.
 	 * @param  array     $config              Array of configuration options for the section.
+	 * @param  array     $custom              Array of additional custom data to be appended to the section.
```
Comment (sixFingers) [inline src/inc/builder/core/api.php]: Allow an extra hash of attributes to be appended to the section.

File: src/inc/builder/core/api.php [hunk 8]
```diff
@@ -195,10 +201,11 @@ function ttfmake_sorter( $a, $b ) {
  * @param  int       $order               The order in which to display the item.
  * @param  string    $path                The path to the template files.
  * @param  array     $config              Array of configuration options for the section.
+ * @param  array     $custom              Array of additional custom data to be appended to the section.
```
Comment (sixFingers) [inline src/inc/builder/core/api.php]: Update the helper to support new section functionality.

File: src/inc/builder/core/base.php [hunk 9]
```diff
@@ -205,19 +209,43 @@ public function display_builder( $post_local ) {
...
+
+			foreach ( $section_order as $section_id ) {
+				array_push( $ordered_sections, $section_data[$section_id] );
+			}
+
+			$section_data = $ordered_sections;
+		}
+
+		// Expose section defaults to JS
+		wp_localize_script( 'ttfmake-builder', 'ttfMakeSectionDefaults', ttfmake_get_section_definitions()->get_section_defaults() );
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Expose a Javascript object with each section's Backbone model default attributes.

File: src/inc/builder/core/base.php [hunk 10]
```diff
@@ -205,19 +209,43 @@ public function display_builder( $post_local ) {
...
+		// Expose saved sections data to JS
+		wp_localize_script( 'ttfmake-builder', 'ttfMakeSectionData', ttfmake_get_section_json_data( $section_data ) );
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Expose a Javascript object with each section's Backbone model saved attributes.

File: src/inc/builder/core/base.php [hunk 11]
```diff
@@ -205,19 +209,43 @@ public function display_builder( $post_local ) {
...
+
+			if ( !is_array( $template ) ) {
+				$templates[$section['id']] = $template;
+			} else {
+				$templates = array_merge( $templates, $template );
+			}
+		}
+
+		// Expose section template strings to JS
+		wp_localize_script( 'ttfmake-builder', 'ttfMakeSectionTemplates', $templates );
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Expose a Javascript object with each section's Underscore template string. This replaces the old section templates script tags.

File: src/inc/builder/core/base.php [hunk 12]
```diff
@@ -457,14 +476,13 @@ function admin_body_class( $classes ) {
 	 * @param  string    $title           Title for the media modal.
 	 * @return string                     Either return the string or echo it.
 	 */
-	public function add_uploader( $section_name, $image_id = 0, $title = '' ) {
+	public function add_uploader( $section_name, $image_id = 0, $title = '', $field_name = 'background-image-url', $field_name_path = '' ) {
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: @tomasstark hints on the docs?

File: src/inc/builder/core/base.php [hunk 13]
```diff
@@ -476,16 +494,18 @@ public function add_uploader( $section_name, $image_id = 0, $title = '' ) {
...
-	 * @param  bool      $iframe           Whether or not to add an iframe to preview content.
+	 * @param  string    $id					The unique ID to identify the different areas.
+	 * @param  string    $textarea_name			The name of the textarea.
+	 * @param  string	 $textarea_name_path	The array / path ot textarea
+	 * @param  string    $content				The content for the text area.
+	 * @param  bool      $iframe				Whether or not to add an iframe to preview content.
 	 * @return void
 	 */
-	public function add_frame( $id, $textarea_name, $content = '', $iframe = true ) {
+	public function add_frame( $id, $textarea_name, $textarea_name_path = '', $content = '', $iframe = true ) {
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: @tomasstark Hints?

File: src/inc/builder/core/base.php [hunk 14]
```diff
@@ -517,9 +537,10 @@ public function add_frame( $id, $textarea_name, $content = '', $iframe = true )
 	 *
 	 * @param  string    $section     The section data.
 	 * @param  array     $data        The data payload to inject into the section.
+	 * @param  boolean   $return      Specifies if the template should be included or returned as string.
 	 * @return void
 	 */
-	public function load_section( $section, $data = array() ) {
+	public function load_section( $section, $data = array(), $return = false ) {
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Allow a section template to be returned as string.

File: src/inc/builder/core/base.php [hunk 15]
```diff
@@ -805,8 +821,9 @@ function ttfmake_load_section_footer() {
  * @param  string    $slug    The relative path and filename (w/out suffix) required to substitute the template in a child theme.
  * @param  string    $path    An optional path extension to point to the template in the parent theme or a plugin.
  * @return string             The template filename if one is located.
+ * @param  boolean   $return  Specifies if the template should be included or returned as string.
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Update the helper to support new section functionality.

File: src/inc/builder/core/base.php [hunk 16]
```diff
@@ -894,7 +913,7 @@ function ttfmake_get_section_name( $data, $is_js_template ) {
 	 * @param array     $data              The section data.
 	 * @param bool      $is_js_template    Whether or not this is in the context of a JS template.
 	 */
-	return apply_filters( 'make_get_section_name', $name, $data, $is_js_template );
+	return apply_filters( 'make_get_section_name', $name, $data, true );
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Section templates in the builder are now Javascript templates by default.

File: src/inc/builder/core/base.php [hunk 17]
```diff
@@ -954,7 +973,7 @@ function ttfmake_get_image_src( $image_id, $size ) {
 		$image = wp_get_attachment_image_src( $image_id, $size );
 
 		if ( false !== $image && isset( $image[0] ) ) {
-			$src = $image;
+			$src = $image[0];
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Return the image url directly, if found.

File: src/inc/builder/core/base.php [hunk 18]
```diff
@@ -1030,39 +1049,32 @@ function ttfmake_register_placeholder_image( $id, $data ) {
...
-	}
-
-	if ( array_key_exists( 'columns-order', $section_data ) && is_array( $section_data['columns-order'] ) ) {
-		$section_data['columns-order'] = implode( ',', $section_data['columns-order'] );
-	}
-
-	if ( array_key_exists( 'banner-slide-order', $section_data ) && is_array( $section_data['banner-slide-order'] ) ) {
-		$section_data['banner-slide-order'] = implode( ',', $section_data['banner-slide-order'] );
-	}
+if ( ! function_exists( 'ttfmake_get_section_json_data' ) ) :
```
Comment (sixFingers) [inline src/inc/builder/core/base.php]: Allow for decorating a section's JSON representation.

File: src/inc/builder/core/base.php (hunk 12 above)
Comment (tomasstark) [inline src/inc/builder/core/base.php]: Docs completed.

**Pull request merged**